Modules/atexitmodule.c

```diff
@@ -276,21 +328,33 @@ static PyObject *
 atexit_unregister(PyObject *module, PyObject *func)
 {
     struct atexit_state *state = get_atexit_state();
+    _PyAtExit_LOCK(state);
     for (int i = 0; i < state->ncallbacks; i++)
     {
         atexit_py_callback *cb = state->callbacks[i];
         if (cb == NULL) {
             continue;
         }
+        PyObject *to_compare = cb->func;
 
-        int eq = PyObject_RichCompareBool(cb->func, func, Py_EQ);
+        // Unlock for custom __eq__ causing re-entrancy
+        _PyAtExit_UNLOCK(state);
+        int eq = PyObject_RichCompareBool(to_compare, func, Py_EQ);
         if (eq < 0) {
             return NULL;
         }
+        _PyAtExit_LOCK(state);
+        if (state->callbacks[i] == NULL)
```

Comment:
The explicit unlocking means that this still has re-entrancy and concurrency issues if the callback list is modified during the comparison. `callbacks[i]` may be a *different* callback.

Using critical sections doesn't completely avoid the problem if `__eq__` is particularly weird, but it avoids it for the common case of identity comparison and the issues are limited to the cases where the GIL-enabled build has issues as well.